### ISSUE TITLE
Nvidia support

### ISSUE THREAD

**Issue by ilya-zlobintsev** [ilya-zlobintsev/LACT#391]
With #388 being merged, LACT now has basic support for Nvidia GPUs through NVML (nvidia management library). This issue tracks the feature support for nvidia.

- [X] Information reporting
  - The UI now only shows fields that have data present, as it doesn't make sense to report vendors-specific info such as CUDA cores or compute units when the value will always be empty on that GPU
- [X] Real-time stats reporting (clockspeed, power usage, power states, fan speed, throttling)
- [X] Power limit configuration
- [X] Custom fan curves
  - Largely uses the same logic as pre-RDNA3 AMD
- [x] Clockspeed configuration
  - Implemented in #398 (only for max values)

Not possible to implement currently:
- Voltage configuration - doesn't appear to be supported in NVML

**Comment by ilya-zlobintsev** [comment]
All the main functionality has been implemented, now it just needs a bit more testing across multiple GPU generations.

**Comment by Dekamir** [comment]
Will we ever be able to toggle individual PowerMizer power/performance levels (or edit them) on NVIDIA?

**Comment by AbdulrahmanObaido** [comment]
can we get intel arc support fan speed set on kernel 6.12 ?

**Comment by ilya-zlobintsev** [comment]
@AbdulrahmanObaido 6.12 only added support for reading the fan speed on ARC, it's not possible to set it (or do any other configuration on intel GPUs)
See #401 for more info

**Comment by Jimmytalksalot** [comment]
what is the config yaml for fan control?

**Comment by stanislav-kozyrev** [comment]
@ilya-zlobintsev Thanks for Nvidia GPU support. By the way, the memory clock offset value in UI and config file is multiplied by 2 compared to GreenWithEnvy and MSI Afterburner. Default memory clock is determined correctly (e.g., 11501 for 23 Gbps). In those programs the offset should be 1500 (AB) instead of 3000 (LACT) to get 26 Gbps on RTX 4080 Super. Please check attached debug snapshot.
[info.json](https://github.com/user-attachments/files/18415398/info.json)

**Comment by stanislav-kozyrev** [comment]
@ilya-zlobintsev Also forgot to mention that sometimes Nvidia drivers don't create `libnvidia-ml.so` (especially beta or from CUDA repo), but there is still `libnvidia-ml.so.1`. I had to make a missing symlink to make `lactd` to detect Nvidia. Maybe it makes sense to fallback to `libnvidia-ml.so.1` if `libnvidia-ml.so` is not found?

**Comment by ilya-zlobintsev** [comment]
@stanislav-kozyrev it was changed to `libnvidia-ml.so.1` since the last stable release, see #414 for more info

**Comment by stanislav-kozyrev** [comment]
@ilya-zlobintsev Thanks for the update. Tried out release 0.7.0 and both issues (library name and memory offset) are resolved.

**Comment by HorstBaerbel** [comment]
Thanks for the nvidia support! RTX 3060 user here. The GPU and VRAM clocks do not seem to adjust correctly. LACT seems to do something to the clocks, but the actual values are a bit off. Also I'd expect the Power states to update to reflect the new values:

![Image](https://github.com/user-attachments/assets/5579a3b4-d4e9-4988-8d63-24485942036d)

[LACT-v0.7.0-snapshot-20250116-163055.tar.gz](https://github.com/user-attachments/files/18441960/LACT-v0.7.0-snapshot-20250116-163055.tar.gz)

**Comment by stanislav-kozyrev** [comment]
@HorstBaerbel Could you please try to reproduce the issue with either GreenWithEnvy (check Flatpacks) or official NVIDIA X Server Settings app? Both apps allow to adjust core and memory offsets, but don't forget to apply defaults with LACT first. Power limit can be changed with `nvidia-smi`. Also, try to get 100% load with GPU heavy benchmark like Unigine Superposition.

The core clock on modern (since last decade or so) GPUs is more like a cap, i.e. upper limit, that a vendor's boost technology aims to achieve given ideal conditions, e.g., hot spot temperature 15C, unlimited power and/or voltage, etc. For instance, out of the box my RTX 4080S core clock is set to 3105 MHz, but in reality it stays around 2655-2750 depending on the load. That's why undervolting modern chips is more beneficial than overclocking -- basically it's all about removing obstacles to let the boost algorithm stretch it's legs.

**Comment by HorstBaerbel** [comment]
@stanislav-kozyrev I'm on Wayland here, so LACT is basically my only hope. GreenWithEnvy and official NVIDIA X Server Settings app don't support Wayland.
About the clocks: I understand that it is an upper limit, but if I increase the limits the clocks do increase but the relation is totally random. This is underclocking with the power limit set to max running FurMark:

![Image](https://github.com/user-attachments/assets/b8a5b931-0b20-48fb-9c22-2451fc639c62)

Not even the reduced-from-default clocks are reached. I'm not saying this is LACTs fault. It may very well be a driver thing.

**Comment by Aspect250** [comment]
Why not implement core clock offset and memory clock offset like in these projects
https://github.com/WickedLukas/nvidia-tuner
https://github.com/Dreaming-Codes/nvidia_oc 
as they mention that they use NVML.
this would provide the ability to overclock and undervolt on nvidia cards.

**Comment by ilya-zlobintsev** [comment]
> Why not implement core clock offset and memory clock offset like in these projects https://github.com/WickedLukas/nvidia-tuner https://github.com/Dreaming-Codes/nvidia_oc as they mention that they use NVML. this would provide the ability to overclock and undervolt on nvidia cards.

The current min/max clock functionality does use offsets under the hood. I plan on showing them as offsets in the ui too, but they already work using the same nvml options.

**Comment by Aspect250** [comment]
> > Why not implement core clock offset and memory clock offset like in these projects https://github.com/WickedLukas/nvidia-tuner https://github.com/Dreaming-Codes/nvidia_oc as they mention that they use NVML. this would provide the ability to overclock and undervolt on nvidia cards.
> 
> The current min/max clock functionality does use offsets under the hood. I plan on showing them as offsets in the ui too, but they already work using the same nvml options.

Im using this [cachyos guide](https://wiki.cachyos.org/configuration/nvidia-wayland-overclock/) as my basis right now, So your using/doing nvmlDeviceSetGpcClkVfOffset according to the guide above and not nvmlDeviceSetGpuLockedClocks or the equivalent to it for the current max gpu clock slider

**Comment by ilya-zlobintsev** [comment]
> > > Why not implement core clock offset and memory clock offset like in these projects https://github.com/WickedLukas/nvidia-tuner https://github.com/Dreaming-Codes/nvidia_oc as they mention that they use NVML. this would provide the ability to overclock and undervolt on nvidia cards.
> > 
> > 
> > The current min/max clock functionality does use offsets under the hood. I plan on showing them as offsets in the ui too, but they already work using the same nvml options.
> 
> Im using this [cachyos guide](https://wiki.cachyos.org/configuration/nvidia-wayland-overclock/) as my basis right now, So your using/doing nvmlDeviceSetGpcClkVfOffset according to the guide above and not nvmlDeviceSetGpuLockedClocks or the equivalent to it for the current max gpu clock slider. Have i got that right?

Yes, that's correct - both the max GPU and VRAM clock settings currently use `nvmlDeviceSetGpcClkVfOffset`/`nvmlDeviceSetMemClkVfOffset`, and the max clock values shown in the GUI are calculated as clockspeed of max pstate + the offset. It was done this way because it allowed to reuse the existing GUI and config options which were originally made for AMD. `nvmlDeviceSetGpuLockedClocks` is not used anywhere.

Here is the relevant code: https://github.com/ilya-zlobintsev/LACT/blob/master/lact-daemon/src/server/gpu_controller/nvidia.rs#L567

However I do plan on replacing the `nvmlDeviceSet*ClkOffset` functions (which are marked as deprecated in NVML docs) with `nvmlDeviceSetClockOffsets`, which allows to configure clock offsets per each pstate separately. It was a bit buggy when I tested it originally, but it appears to be improved with the 570 beta driver.

**Comment by Aspect250** [comment]
Thanks for clarifying. So much easier to undervolt/overclock right now.

Cheers for adding NVIDIA support.

**Comment by HorstBaerbel** [comment]
Only here to report that this works much better now with 570.86.16. Thanks!
